crates/client/gateway/src/server/handler.rs

```diff
@@ -0,0 +1,307 @@
+        Ok(block_id) => block_id,
+        // Return the error response if the request is malformed
+        Err(e) => return e.into(),
+    };
+
+    let resolved_block_id = match backend
+        .resolve_block_id(&block_id)
+        .or_internal_server_error("resolve_block_id")
+        .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+    {
+        Ok(block_id) => block_id,
+        Err(e) => return e.into(),
+    };
+
+    let state_diff = match backend
+        .get_block_state_diff(&resolved_block_id)
+        .or_internal_server_error("get_block_state_diff")
+        .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+    {
+        Ok(state_diff) => state_diff,
+        Err(e) => return e.into(),
+    };
+
+    let with_block = if include_block_params(&params) {
+        let block = match backend
+            .get_block(&block_id)
+            .or_internal_server_error("get_block")
+            .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+        {
+            Ok(block) => block,
+            Err(e) => return e.into(),
+        };
+        Some(block)
+    } else {
+        None
+    };
+
+    match resolved_block_id.is_pending() {
+        true => {
+            let old_root = match backend
+                .get_block_info(&BlockId::Tag(BlockTag::Latest))
+                .or_internal_server_error("get_block_state_diff")
+            {
+                Ok(Some(block)) => match block.as_nonpending() {
+                    Some(block) => block.header.global_state_root,
+                    None => return GatewayError::InternalServerError.into(),
+                },
+                Ok(None) => Felt::ZERO, // The pending block is actually genesis, so old root is zero
+                Err(e) => return e.into(),
+            };
+            let state_update = PendingStateUpdateProvider { old_root, state_diff: state_diff.into() };
+
+            if let Some(block) = with_block {
+                let block = match MadaraPendingBlock::try_from(block.clone()) {
+                    Ok(block) => block,
+                    Err(_) => return GatewayError::InternalServerError.into(),
+                };
+                let block_provider = PendingBlockProvider::new(block);
+                create_json_response(
+                    hyper::StatusCode::OK,
+                    &json!({"block": block_provider, "state_update": state_update}),
+                )
+            } else {
+                create_json_response(hyper::StatusCode::OK, &state_update)
+            }
+        }
+        false => {
+            let block_info = match backend
+                .get_block_info(&resolved_block_id)
+                .or_internal_server_error("get_block_info")
+                .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+            {
+                Ok(block_info) => block_info,
+                Err(e) => return e.into(),
```

Comment:
what does that into do?
this whole match can just be replaced with the question mark operator